db/c.cc

```diff
@@ -2631,6 +2631,16 @@ rocksdb_iterator_t* rocksdb_writebatch_wi_create_iterator_with_base(
   return result;
 }
 
+rocksdb_iterator_t* rocksdb_writebatch_wi_create_iterator_with_base_readopts(
+    rocksdb_writebatch_wi_t* wbwi, rocksdb_iterator_t* base_iterator,
+    const rocksdb_readoptions_t* options) {
+  rocksdb_iterator_t* result = new rocksdb_iterator_t;
+  result->rep =
+      wbwi->rep->NewIteratorWithBase(base_iterator->rep, &options->rep);
+  delete base_iterator;
```

Comment:
I suspect this might be a bug - the `NewIteratorWithBase` doc comment says:

> The returned iterator should be deleted by the caller.
> The base_iterator is now 'owned' by the returned iterator. Deleting the
> returned iterator will also delete the base_iterator.

And indeed if you look at the `BaseDeltaIterator` class, the `base_iterator` parameter is stashed into a `std::unique_ptr` field in, which will own deallocation on destruction. I realise the pre-existing `rocksdb_writebatch_wi_create_iterator_with_base` / `rocksdb_writebatch_wi_create_iterator_with_base_cf` methods do this too, and there is test coverage for the former in `c_test.c`, so it should have blown up if there was a double-free. I'd like to dig a bit deeper.

Reply:
I think this is fine because `rocksdb_iterator_t` is defined as

```
struct rocksdb_iterator_t {
  Iterator* rep;
};
```

Hence, deleting this type will only delete the struct but not the pointed to inner iterator. You can also see it here:

```
void rocksdb_iter_destroy(rocksdb_iterator_t* iter) {
  delete iter->rep;
  delete iter;
}
```

Reply:
I ran a test in the debugger just to be sure, and I can confirm that the base iterator gets destroyed correctly by way of `rocksdb_iter_destroy` -> `~BaseDeltaIterator` without the explicit `delete base_iterator` here. I'm not sure what that says about why the code works as is 😅 I'd love for someone to enlighten me!

Reply:
Do I understand you correctly that `delete base_iterator` (on line 2640) with `rocksdb_iter_destroy* base_iterator` also runs the `~BaseDeltaIterator` destructor?

Reply:
Thanks for helping me clear my misunderstanding! On this line we're only deleting the pointer wrapper of type `rocksdb_iterator_t`, not the underlying base iterator. The fact that I didn't see the destructor `~BaseDeltaIterator` get called when stepping over this line in the debugger should have given me a hint.

```
struct rocksdb_iterator_t {
  Iterator* rep;
};
```

All good, going to resolve this!